lib/free_quota.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+require "yaml"
+
+class FreeQuota
+  # :nocov:
+  def self.freeze
+    free_quotas
+    super
+  end
+  # :nocov:
+
+  def self.free_quotas
+    @free_quotas ||= begin
+      quotas = YAML.load_file("config/free_quotas.yml")
+      quotas.each_with_object({}) do |item, hash|
+        hash[item["name"]] = item
+      end
+    end
+    @free_quotas
+  end
+
+  def self.get_applicable_billing_rate_ids(free_quota)
+    BillingRate.from_resource_type(free_quota["resource_type"]).map { _1["id"] }
+  end
+
+  def self.remaining_free_quota(name, project_id)
+    free_quota = free_quotas[name]
+    begin_time = get_begin_time
+    tokens_used = BillingRecord.where(
+      project_id: project_id,
+      billing_rate_id: get_applicable_billing_rate_ids(free_quota)
+    ).where {
+      Sequel.pg_range(span).overlaps(Sequel.pg_range(begin_time...Time.now))
+    }.sum(:amount) || 0
+    [0, free_quota["value"] - tokens_used].max
+  end
```

Comment:
nit:
- not need to `begin_time` since it's used once. we might also rename the method as `begin_time` instead of `get_begin_time`
- single line `where` clauses look more like the `self.get_free_quota_exhausted_projects` method.
- if you apply https://github.com/ubicloud/ubicloud/pull/2818/files#r1956158740
- `tokens_used` -> `used_amount`
- `project_id: project_id` -> `project_id:`

```suggestion
  def self.remaining_free_quota(name, project_id)
    free_quota = free_quotas[name]
    used_amount = BillingRecord
      .where(project_id:, billing_rate_id: free_quota["billing_rate_ids"])
      .where { Sequel.pg_range(span).overlaps(Sequel.pg_range(begin_time...Time.now)) }
      .sum(:amount) || 0
    [0, free_quota["value"] - used_amount].max
  end
```

Reply:
Done.